File: Labs/Lab6/problem.java
import java.util.LinkedList;
import java.util.Queue;

public class CourseScheduler {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] inDegree = new int[numCourses];
        LinkedList<Integer>[] graph = new LinkedList[numCourses];
        for (int i = 0; i < numCourses; i++) {
            graph[i] = new LinkedList<>();
        }
        for (int[] pair : prerequisites) {
            inDegree[pair[0]]++;
            graph[pair[1]].add(pair[0]);
        }
        
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (inDegree[i] == 0) {
                queue.add(i);
            }
        }
        
        int count = 0;
        while (!queue.isEmpty()) {
            int course = queue.poll();
            count++;
            for (int nextCourse : graph[course]) {
                inDegree[nextCourse]--;
                if (inDegree[nextCourse] == 0) {
                    queue.add(nextCourse);
                }
            }
        }
        
        return count == numCourses;
    }

    public static void main(String[] args) {
        CourseScheduler scheduler = new CourseScheduler();
        System.out.println(scheduler.canFinish(2, new int[][]{{1, 0}})); // true
        System.out.println(scheduler.canFinish(2, new int[][]{{1, 0}, {0, 1}})); // false
    }
}

File: Labs/Lab6/problem.py
def canFinish(numCourses, prerequisites):
    graph = [[] for _ in range(numCourses)]
    visited = [0] * numCourses  # 0: unvisited, 1: visiting, -1: visited

    def dfs(course):
        if visited[course] == 1:
            return False
        if visited[course] == -1:
            return True
        visited[course] = 1
        for next_course in graph[course]:
            if not dfs(next_course):
                return False
        visited[course] = -1
        return True

    for dest, src in prerequisites:
        graph[src].append(dest)

    for course in range(numCourses):
        if visited[course] == 0:
            if not dfs(course):
                return False
    return True

# Testing the function
print(canFinish(2, [[1, 0]]))  # True
print(canFinish(2, [[1, 0], [0, 1]]))  # False
